Use single stream for apply optimizer in backward

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #89963
* #89194

Per title

Differential Revision: [D41622106](https://our.internmc.facebook.com/intern/diff/D41622106/)

rohan-varma committed:
Use single stream for apply optimizer in backward

Per title

Differential Revision: [D41622106](https://our.internmc.facebook.com/intern/diff/D41622106/)

[ghstack-poisoned]

zhaojuanmao commented on torch/distributed/algorithms/ddp_comm_hooks/optimizer_overlap_hooks.py:
```diff
@@ -40,23 +49,55 @@ def _apply_optim_in_backward_hook(
-        return fut.then(optimizer_step)
+        # enqueue a callback to wait for this optimizer stream at the end of
+        # backward.
+        def wait_for_optim_stream_callback():
+            torch.cuda.current_stream().wait_stream(
+                optim_stream_state.optim_stream
+            )
+            # reset for the next backwards pass
+            optim_stream_state.wait_for_optim_stream_enqueued = False
+
+        if not optim_stream_state.wait_for_optim_stream_enqueued:
+            Variable._execution_engine.queue_callback(
+                wait_for_optim_stream_callback
+            )
+            # mark that the callback is enqueued
+            optim_stream_state.wait_for_optim_stream_enqueued = True
+
+        return fut
```
we do not need to queue call back, instead we can simply call " torch.cuda.current_stream().wait_stream(
                optim_stream_state.optim_stream)" in the final_backward call back?

rohan-varma replied:
Is this referring to calling this in `finalize_backward` in C++? I agree we can do this, I am just not sure if we can pass a cuda stream from python to C++, but it should be possible.

zhaojuanmao reviewed (approved):
left one comment

rohan-varma committed:
Update on "Use single stream for apply optimizer in backward"


Per title

Differential Revision: [D41622106](https://our.internmc.facebook.com/intern/diff/D41622106/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Use single stream for apply optimizer in backward"


Per title

Differential Revision: [D41622106](https://our.internmc.facebook.com/intern/diff/D41622106/)

[ghstack-poisoned]

rohan-varma commented:
close in favor of https://github.com/pytorch/pytorch/pull/89194
